### ISSUE TITLE
More multi-threading for vg pack when using a GAF input

### ISSUE THREAD

**Issue by jmonlong** [vgteam/vg#3007]
In my experience, vg pack uses maximum 4 threads when the input is GAF. 

If there is a way to use more threads that would help me fully use the cloud instances that I run for this part of my analysis. I just realized that because I need to use a specific amount of memory, the instance type gets bumped up to 16 cores even if I ask for say 4 cores, so we end up paying for unused cores.

This is not urgent because I can always switch back to GAM.

**Comment by ekg** [comment]
A standalone tool to do this could be very useful.

It might be IO bound. Is that possible?

**Comment by glennhickey** [comment]
I'm not sure what's going on, and got sidetracked before figuring it out the last couple times I looked.  

`pack` is not doing much with each read, so I think I/O could be part of it.  But at the same time, GAF is slower single threaded than GAM.  There's going to be an overhead parsing a string then filling a protobuf object vs deserializing the protobuf (which is also what kcachegrind's telling me), but it seems bigger than what I expect in `pack`.  I think moving to lower-level API for the string stuff outght to help, but hasn't made much difference in the (admittedly small) tests I've run so far.  That said, you'd think that this overhead would let GAF use more threads than GAM, not fewer.

**Comment by glennhickey** [comment]
Working on a small test of 3M single-threaded GAM records:
```
Load the GAM into memory: 5.4s
Load the uncompressed GAF into GAF records: 5.6s
Load the GAF into GAM records: 21s
```
If callgrind's to judge, that 15s overhead is mostly xg lookups.  Parsing the `cs-cigar` into edits requires a lot of node lengths.  Doing the `sequence` fields (querying dna strings from the graph) seems to take the most time though.  `pack` doesn't need sequence, so there's promise in just skipping them.  Also, using non-xg formats speeds up sequence access (21s for xg becomes 19s, 18s and 12s for packed-graph, hash-graph and odgi, respectively), but would probably slow down the rest of pack.

Anyway, will start some bigger multithreaded tests.

**Comment by glennhickey** [comment]
OK, I think this should be fixed finally in master by way of #3008, which parallelizes the internal GAF->GAM conversion that seemed to be the bottleneck in my tests.

**Comment by jmonlong** [comment]
Awesome, thanks Glenn!
